tools/asynccmd/filterbuffer.go

```diff
@@ -0,0 +1,326 @@
+				if indexes == nil {
+					indexes = []int{}
+				}
+				partialMatchMap[secretIdx] = append(indexes, lineIdx)
+			}
+		}
+	}
+
+	return matchMap, partialMatchMap
+}
+
+// linesToKeepRange returns a range (first, last index) of lines needs to be observed
+// since they contain partially matching secrets
+func (b *Buffer) linesToKeepRange(partialMatchMap map[int][]int) (int, int) {
+	first := -1
+	last := -1
+
+	for secretIdx, lineIndexes := range partialMatchMap {
+		secret := b.secrets[secretIdx]
+		secretLength := len(secret)
+
+		for _, lineIdx := range lineIndexes {
+			if first == -1 {
+				first = lineIdx
+				last = first + secretLength
+				continue
+			}
+
+			if lineIdx+secretLength > last {
+				last = lineIdx + secretLength
+			}
+
+			if first > lineIdx {
+				first = lineIdx
+			}
+		}
+	}
+
+	return first, last
+}
+
+// matchLines return which lines can be printed and which should be keept for further observing
+func (b *Buffer) matchLines(lines [][]byte, partialMatchMap map[int][]int) ([][]byte, [][]byte) {
+	first, last := b.linesToKeepRange(partialMatchMap)
+	if first == -1 {
+		return lines[:], [][]byte{}
+	}
+
+	if first == 0 {
+		if last > len(lines)-1 {
+			return [][]byte{}, lines[0:len(lines)]
+		}
+		return lines[last:], lines[0:last]
+	}
+
+	if last > len(lines)-1 {
+		return lines[:first], lines[first:len(lines)]
+	}
+	return lines[:first], lines[first:last]
+}
+
+// redact hides the given secret from the lines
+func (b *Buffer) redact(lines [][]byte, matchMap map[int][]int) [][]byte {
+	redacted := lines[:]
+	for secretIdx, lineIndexes := range matchMap {
+		secret := b.secrets[secretIdx]
+
+		for _, lineIdx := range lineIndexes {
+			if lineIdx > len(lines)-1 {
+				continue
+			}
+
+			for i := lineIdx; i < lineIdx+len(secret); i++ {
+				secretLine := secret[i-lineIdx]
+				line := redacted[i]
+				redacted[i] = bytes.Replace(line, secretLine, []byte(RedactStr), -1)
+			}
+		}
+	}
+	return redacted
+}
+
+// secretsByteList returns the list of secret byte lines
+func secretsByteList(secrets []string) [][][]byte {
+	secretByteLinesList := [][][]byte{}
+	for _, secret := range secrets {
+		secretBytes := []byte(secret)
+		secretByteLines := bytes.Split(secretBytes, newLine)
+		secretByteLinesList = append(secretByteLinesList, secretByteLines)
+	}
+	return secretByteLinesList
+}
+
+// split p after "\n", the split is assigned to lines
+// if last line has no "\n" it is assigned to chunk
+func split(p []byte) (lines [][]byte, chunk []byte) {
+	if p == nil || len(p) == 0 {
+		return [][]byte{}, []byte{}
+	}
+
+	lines = [][]byte{}
+	chunk = p[:]
+	for len(chunk) > 0 {
+		idx := bytes.Index(chunk, newLine)
+		if idx == -1 {
+			return
+		}
+
+		lines = append(lines, chunk[:idx+1])
+
+		if idx == len(chunk)-1 {
+			chunk = []byte{}
+		} else {
+			chunk = chunk[idx+1:]
+		}
+	}
+	return
+}
+
+// join lines and chunk to restore p after split(p)
+func join(lines [][]byte) []byte {
+	if lines == nil || len(lines) == 0 {
```

Comment:
no need for these checks. If `len(lines) == 0` then the `bytes.Join` function will return with `[]byte{}`.